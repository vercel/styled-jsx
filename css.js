function notTranspiledError(name) {
  throw new Error(
    'styled-jsx/css: if you are getting this error it means that your `' +
      name +
      '` tagged template literals were not transpiled.'
  )
}
function css() {
  notTranspiledError('css')
}
css.global = function() {
  notTranspiledError('global')
}
css.resolve = function() {
  notTranspiledError('resolve')
}
module.exports = css
module.exports.global = css.global
module.exports.resolve = css.resolve
