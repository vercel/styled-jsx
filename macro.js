module.exports = require('./dist/macro')
