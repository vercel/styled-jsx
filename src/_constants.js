export const STYLE_ATTRIBUTE = 'jsx'
export const GLOBAL_ATTRIBUTE = 'global'
export const STYLE_COMPONENT_ID = 'styleId'
export const MARKUP_ATTRIBUTE = 'data-jsx'
export const MARKUP_ATTRIBUTE_EXTERNAL = 'data-jsx-ext'
export const STYLE_COMPONENT = '_JSXStyle'
export const STYLE_COMPONENT_CSS = 'css'
