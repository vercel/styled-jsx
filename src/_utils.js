import * as t from 'babel-types'
import _hashString from 'string-hash'
import { SourceMapGenerator } from 'source-map'
import convert from 'convert-source-map'
import transform from './lib/style-transform'

import {
  STYLE_ATTRIBUTE,
  GLOBAL_ATTRIBUTE,
  STYLE_COMPONENT_ID,
  STYLE_COMPONENT,
  STYLE_COMPONENT_CSS,
  STYLE_COMPONENT_DYNAMIC
} from './_constants'

const concat = (a, b) => t.binaryExpression('+', a, b)
const and = (a, b) => t.logicalExpression('&&', a, b)
const or = (a, b) => t.logicalExpression('||', a, b)

const joinSpreads = spreads => spreads.reduce((acc, curr) => or(acc, curr))

export const hashString = str => String(_hashString(str))

export const addClassName = (path, jsxId) => {
  const jsxIdWithSpace = concat(jsxId, t.stringLiteral(' '))
  const attributes = path.get('attributes')
  const spreads = []
  let className = null
  // Find className and collect spreads
  for (let i = attributes.length - 1, attr; (attr = attributes[i]); i--) {
    const node = attr.node
    if (t.isJSXSpreadAttribute(attr)) {
      const name = node.argument.name
      const attrNameDotClassName = t.memberExpression(
        t.isMemberExpression(node.argument)
          ? node.argument
          : t.identifier(name),
        t.identifier('className')
      )

      spreads.push(
        // `${name}.className != null && ${name}.className`
        and(
          t.binaryExpression('!=', attrNameDotClassName, t.nullLiteral()),
          attrNameDotClassName
        )
      )
      continue
    }
    if (t.isJSXAttribute(attr) && node.name.name === 'className') {
      className = attributes[i]
      // found className break the loop
      break
    }
  }

  if (className) {
    let newClassName = className.node.value.expression || className.node.value
    newClassName =
      t.isStringLiteral(newClassName) || t.isTemplateLiteral(newClassName)
        ? newClassName
        : or(newClassName, t.stringLiteral(''))
    className.remove()

    className = t.jSXExpressionContainer(
      spreads.length === 0
        ? concat(jsxIdWithSpace, newClassName)
        : concat(jsxIdWithSpace, or(joinSpreads(spreads), newClassName))
    )
  } else {
    className = t.jSXExpressionContainer(
      spreads.length === 0
        ? jsxId
        : concat(jsxIdWithSpace, or(joinSpreads(spreads), t.stringLiteral('')))
    )
  }

  path.node.attributes.push(
    t.jSXAttribute(t.jSXIdentifier('className'), className)
  )
}

export const getScope = path =>
  (
    path.findParent(
      path =>
        path.isFunctionDeclaration() ||
        path.isArrowFunctionExpression() ||
        path.isClassMethod()
    ) || path
  ).scope

export const isGlobalEl = el =>
  el.attributes.some(({ name }) => name && name.name === GLOBAL_ATTRIBUTE)

export const isStyledJsx = ({ node: el }) =>
  t.isJSXElement(el) &&
  el.openingElement.name.name === 'style' &&
  el.openingElement.attributes.some(attr => attr.name.name === STYLE_ATTRIBUTE)

export const findStyles = path => {
  if (isStyledJsx(path)) {
    const { node } = path
    return isGlobalEl(node.openingElement) ? [path] : []
  }

  return path.get('children').filter(isStyledJsx)
}

// The following visitor ensures that MemberExpressions and Identifiers
// are not in the scope of the current Method (render) or function (Component).
export const validateExpressionVisitor = {
  MemberExpression(path, scope) {
    const { node } = path
    if (
      (t.isIdentifier(node.property) &&
        (t.isThisExpression(node.object) &&
          (node.property.name === 'props' ||
            node.property.name === 'state' ||
            node.property.name === 'context'))) ||
      (t.isIdentifier(node.object) && scope.hasOwnBinding(node.object.name))
    ) {
      throw path.buildCodeFrameError(
        `Expected a constant ` +
          `as part of the template literal expression ` +
          `(eg: <style jsx>{\`p { color: $\{myColor}\`}</style>), ` +
          `but got a MemberExpression: this.${node.property.name}`
      )
    }
  },
  Identifier(path, scope) {
    const { name } = path.node

    if (t.isMemberExpression(path.parentPath) && scope.hasOwnBinding(name)) {
      return
    }

    let targetScope = path.scope
    let isDynamicBinding = false

    // Traversing scope chain in order to find current variable.
    // If variable has no parent scope and it's `const` then we can interp. it
    // as static in order to optimize styles.
    // `let` and `var` can be changed during runtime.
    while (targetScope) {
      if (targetScope.hasOwnBinding(name)) {
        const binding = targetScope.bindings[name]
        isDynamicBinding =
          binding.scope.parent !== null || binding.kind !== 'const'
        break
      }
      targetScope = targetScope.parent
    }

    if (isDynamicBinding) {
      throw path.buildCodeFrameError(
        `Expected \`${name}\` ` +
          `to not come from the closest scope.\n` +
          `Styled JSX encourages the use of constants ` +
          `instead of \`props\` or dynamic values ` +
          `which are better set via inline styles or \`className\` toggling. ` +
          `See https://github.com/zeit/styled-jsx#dynamic-styles`
      )
    }
  }
}

// Use `validateExpressionVisitor` to determine whether the `expr`ession has dynamic values.
export const isDynamic = (expr, scope) => {
  try {
    expr.traverse(validateExpressionVisitor, scope)
    return false
  } catch (err) {}

  return true
}

const validateExternalExpressionsVisitor = {
  Identifier(path) {
    if (t.isMemberExpression(path.parentPath)) {
      return
    }
    const { name } = path.node
    if (!path.scope.hasBinding(name)) {
      throw path.buildCodeFrameError(path.getSource())
    }
  },
  MemberExpression(path) {
    const { node } = path
    if (!t.isIdentifier(node.object)) {
      return
    }
    if (!path.scope.hasBinding(node.object.name)) {
      throw path.buildCodeFrameError(path.getSource())
    }
  },
  ThisExpression(path) {
    throw new Error(path.parentPath.getSource())
  }
}

export const validateExternalExpressions = path => {
  try {
    path.traverse(validateExternalExpressionsVisitor)
  } catch (err) {
    throw path.buildCodeFrameError(`
      Found an \`undefined\` or invalid value in your styles: \`${
        err.message
      }\`.

      If you are trying to use dynamic styles in external files this is unfortunately not possible yet.
      Please put the dynamic parts alongside the component. E.g.

      <button>
        <style jsx>{externalStylesReference}</style>
        <style jsx>{\`
          button { background-color: $\{${err.message}} }
        \`}</style>
      </button>
    `)
  }
}

export const getJSXStyleInfo = (expr, scope) => {
  const { node } = expr
  const location = node.loc

  // Assume string literal
  if (t.isStringLiteral(node)) {
    return {
      hash: hashString(node.value),
      css: node.value,
      expressions: [],
      dynamic: false,
      location
    }
  }

  // Simple template literal without expressions
  if (node.expressions.length === 0) {
    return {
      hash: hashString(node.quasis[0].value.cooked),
      css: node.quasis[0].value.cooked,
      expressions: [],
      dynamic: false,
      location
    }
  }

  // Special treatment for template literals that contain expressions:
  //
  // Expressions are replaced with a placeholder
  // so that the CSS compiler can parse and
  // transform the css source string
  // without having to know about js literal expressions.
  // Later expressions are restored.
  //
  // e.g.
  // p { color: ${myConstant}; }
  // becomes
  // p { color: %%styled-jsx-placeholder-${id}%%; }

  const { quasis, expressions } = node
  const hash = hashString(expr.getSource().slice(1, -1))
  const dynamic = scope ? isDynamic(expr, scope) : false
  const css = quasis.reduce(
    (css, quasi, index) =>
      `${css}${quasi.value.cooked}${
        quasis.length === index + 1 ? '' : `%%styled-jsx-placeholder-${index}%%`
      }`,
    ''
  )

  return {
    hash,
    css,
    expressions,
    dynamic,
    location
  }
}

export const computeClassNames = (styles, externalJsxId) => {
  if (styles.length === 0) {
    return {
      className: externalJsxId
    }
  }

  const hashes = styles.reduce(
    (acc, styles) => {
      if (styles.dynamic === false) {
        acc.static.push(styles.hash)
      } else {
        acc.dynamic.push(styles)
      }
      return acc
    },
    {
      static: [],
      dynamic: []
    }
  )

  const staticClassName = `jsx-${hashString(hashes.static.join(','))}`

  // Static and optionally external classes. E.g.
  // '[jsx-externalClasses] jsx-staticClasses'
  if (hashes.dynamic.length === 0) {
    return {
      staticClassName,
      className: externalJsxId
        ? concat(t.stringLiteral(staticClassName + ' '), externalJsxId)
        : t.stringLiteral(staticClassName)
    }
  }

  // _JSXStyle.dynamic([ ['1234', [props.foo, bar, fn(props)]], ... ])
  const dynamic = t.callExpression(
    // Callee: _JSXStyle.dynamic
    t.memberExpression(t.identifier(STYLE_COMPONENT), t.identifier('dynamic')),
    // Arguments
    [
      t.arrayExpression(
        hashes.dynamic.map(styles =>
          t.arrayExpression([
            t.stringLiteral(hashString(styles.hash + staticClassName)),
            t.arrayExpression(styles.expressions)
          ])
        )
      )
    ]
  )

  // Dynamic and optionally external classes. E.g.
  // '[jsx-externalClasses] ' + _JSXStyle.dynamic([ ['1234', [props.foo, bar, fn(props)]], ... ])
  if (hashes.static.length === 0) {
    return {
      staticClassName,
      className: externalJsxId
        ? concat(concat(externalJsxId, t.stringLiteral(' ')), dynamic)
        : dynamic
    }
  }

  // Static, dynamic and optionally external classes. E.g.
  // '[jsx-externalClasses] jsx-staticClasses ' + _JSXStyle.dynamic([ ['5678', [props.foo, bar, fn(props)]], ... ])
  return {
    staticClassName,
    className: externalJsxId
      ? concat(
          concat(externalJsxId, t.stringLiteral(` ${staticClassName} `)),
          dynamic
        )
      : concat(t.stringLiteral(`${staticClassName} `), dynamic)
  }
}

export const templateLiteralFromPreprocessedCss = (css, expressions) => {
  const quasis = []
  const finalExpressions = []
  const parts = css.split(/(?:%%styled-jsx-placeholder-(\d+)%%)/g)

  if (parts.length === 1) {
    return t.stringLiteral(css)
  }

  parts.forEach((part, index) => {
    if (index % 2 > 0) {
      // This is necessary because, after preprocessing, declarations might have been alterate.
      // eg. properties are auto prefixed and therefore expressions need to match.
      finalExpressions.push(expressions[part])
    } else {
      quasis.push(part)
    }
  })

  return t.templateLiteral(
    quasis.map((quasi, index) =>
      t.templateElement(
        {
          raw: quasi,
          cooked: quasi
        },
        quasis.length === index + 1
      )
    ),
    finalExpressions
  )
}

export const cssToBabelType = css => {
  if (typeof css === 'string') {
    return t.stringLiteral(css)
  }

  if (Array.isArray(css)) {
    return t.arrayExpression(css)
  }

  return css
}

export const makeStyledJsxTag = (id, transformedCss, expressions = []) => {
  const css = cssToBabelType(transformedCss)

  const attributes = [
    t.jSXAttribute(
      t.jSXIdentifier(STYLE_COMPONENT_ID),
      t.jSXExpressionContainer(
        typeof id === 'string' ? t.stringLiteral(id) : id
      )
    ),
    t.jSXAttribute(
      t.jSXIdentifier(STYLE_COMPONENT_CSS),
      t.jSXExpressionContainer(css)
    )
  ]

  if (expressions.length > 0) {
    attributes.push(
      t.jSXAttribute(
        t.jSXIdentifier(STYLE_COMPONENT_DYNAMIC),
        t.jSXExpressionContainer(t.arrayExpression(expressions))
      )
    )
  }

  return t.jSXElement(
    t.jSXOpeningElement(t.jSXIdentifier(STYLE_COMPONENT), attributes, true),
    null,
    []
  )
}

export const makeSourceMapGenerator = file => {
  const filename = file.opts.sourceFileName
  const generator = new SourceMapGenerator({
    file: filename,
    sourceRoot: file.opts.sourceRoot
  })

  generator.setSourceContent(filename, file.code)
  return generator
}

export const addSourceMaps = (code, generator, filename) => {
  const sourceMaps = [
    convert.fromObject(generator).toComment({ multiline: true }),
    `/*@ sourceURL=${filename} */`
  ]

  if (Array.isArray(code)) {
    return code.concat(sourceMaps)
  }

  return [code].concat(sourceMaps).join('\n')
}

export const combinePlugins = plugins => {
  if (!plugins) {
    return css => css
  }

  if (
    !Array.isArray(plugins) ||
    plugins.some(p => !Array.isArray(p) && typeof p !== 'string')
  ) {
    throw new Error(
      '`plugins` must be an array of plugins names (string) or an array `[plugin-name, {options}]`'
    )
  }

  return plugins
    .map((plugin, i) => {
      let options = {}
      if (Array.isArray(plugin)) {
        options = plugin[1] || {}
        plugin = plugin[0]
        if (Object.prototype.hasOwnProperty.call(options, 'babel')) {
          throw new Error(`
            Error while trying to register the styled-jsx plugin: ${plugin}
            The option name \`babel\` is reserved.
          `)
        }
      }

      // eslint-disable-next-line import/no-dynamic-require
      let p = require(plugin)
      if (p.default) {
        p = p.default
      }

      const type = typeof p
      if (type !== 'function') {
        throw new Error(
          `Expected plugin ${
            plugins[i]
          } to be a function but instead got ${type}`
        )
      }
      return {
        plugin: p,
        options
      }
    })
    .reduce(
      (previous, { plugin, options }) => (css, babelOptions) =>
        plugin(previous ? previous(css, babelOptions) : css, {
          ...options,
          babel: babelOptions
        }),
      null
    )
}

const getPrefix = (isDynamic, id) =>
  isDynamic ? '.__jsx-style-dynamic-selector' : `.${id}`

export const processCss = (stylesInfo, options) => {
  const {
    hash,
    css,
    expressions,
    dynamic,
    location,
    fileInfo,
    isGlobal,
    plugins,
    vendorPrefixes
  } = stylesInfo

  const staticClassName =
    stylesInfo.staticClassName || `jsx-${hashString(hash)}`

  const { splitRules } = options

  const useSourceMaps = Boolean(fileInfo.sourceMaps) && !splitRules

  const pluginsOptions = {
    location: {
      start: { ...location.start },
      end: { ...location.end }
    },
    vendorPrefixes,
    sourceMaps: useSourceMaps,
    isGlobal,
    filename: fileInfo.filename
  }

  let transformedCss

  if (useSourceMaps) {
    const generator = makeSourceMapGenerator(fileInfo.file)
    const filename = fileInfo.sourceFileName
    transformedCss = addSourceMaps(
      transform(
        isGlobal ? '' : getPrefix(dynamic, staticClassName),
        plugins(css, pluginsOptions),
        {
          generator,
          offset: location.start,
          filename,
          splitRules,
          vendorPrefixes
        }
      ),
      generator,
      filename
    )
  } else {
    transformedCss = transform(
      isGlobal ? '' : getPrefix(dynamic, staticClassName),
      plugins(css, pluginsOptions),
      { splitRules, vendorPrefixes }
    )
  }

  if (expressions.length > 0) {
    if (typeof transformedCss === 'string') {
      transformedCss = templateLiteralFromPreprocessedCss(
        transformedCss,
        expressions
      )
    } else {
      transformedCss = transformedCss.map(transformedCss =>
        templateLiteralFromPreprocessedCss(transformedCss, expressions)
      )
    }
  } else if (Array.isArray(transformedCss)) {
    transformedCss = transformedCss.map(transformedCss =>
      t.stringLiteral(transformedCss)
    )
  }

  return {
    hash: dynamic ? hashString(hash + staticClassName) : hashString(hash),
    css: transformedCss,
    expressions: dynamic && expressions
  }
}

export const booleanOption = opts => {
  let ret
  opts.some(opt => {
    if (typeof opt === 'boolean') {
      ret = opt
      return true
    }
    return false
  })
  return ret
}
