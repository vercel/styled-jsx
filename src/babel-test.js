import jsx from 'babel-plugin-syntax-jsx'

export default function() {
  return {
    inherits: jsx,
    visitor: {
      JSXOpeningElement(path) {
        const el = path.node
        const { name } = el.name || {}

        if (name !== 'style') {
          return
        }

        el.attributes = el.attributes.filter(a => a.name.name !== 'jsx')
      }
    }
  }
}
