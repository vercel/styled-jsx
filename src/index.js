export {
  StyleRegistry,
  createStyleRegistry,
  useStyleRegistry
} from './stylesheet-registry'
