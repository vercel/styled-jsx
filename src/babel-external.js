import * as t from 'babel-types'

import { STYLE_COMPONENT } from './_constants'

import {
  getJSXStyleInfo,
  processCss,
  cssToBabelType,
  validateExternalExpressions,
  getScope,
  computeClassNames,
  makeStyledJsxTag,
  createReactComponentImportDeclaration,
  setStateOptions
} from './_utils'

const isModuleExports = t.buildMatchMemberExpression('module.exports')

export function processTaggedTemplateExpression({
  type,
  path,
  fileInfo,
  splitRules,
  plugins,
  vendorPrefix
}) {
  const templateLiteral = path.get('quasi')
  let scope

  // Check whether there are undefined references or
  // references to this.something (e.g. props or state).
  // We allow dynamic styles only when resolving styles.
  if (type !== 'resolve') {
    validateExternalExpressions(templateLiteral)
  } else if (!path.scope.path.isProgram()) {
    scope = getScope(path)
  }

  const stylesInfo = getJSXStyleInfo(templateLiteral, scope)

  const { staticClassName, className } = computeClassNames([stylesInfo])

  const styles = processCss(
    {
      ...stylesInfo,
      staticClassName,
      fileInfo,
      isGlobal: type === 'global',
      plugins,
      vendorPrefix
    },
    { splitRules }
  )

  if (type === 'resolve') {
    const { hash, css, expressions } = styles
    path.replaceWith(
      // {
      //   styles: <_JSXStyle ... />,
      //   className: 'jsx-123'
      // }
      t.objectExpression([
        t.objectProperty(
          t.identifier('styles'),
          makeStyledJsxTag(hash, css, expressions)
        ),
        t.objectProperty(t.identifier('className'), className)
      ])
    )
    return
  }

  const id = path.parentPath.node.id
  const baseExportName = id ? id.name : 'default'
  let parentPath =
    baseExportName === 'default'
      ? path.parentPath
      : path.findParent(
          path =>
            path.isVariableDeclaration() ||
            (path.isAssignmentExpression() &&
              isModuleExports(path.get('left').node))
        )

  if (baseExportName !== 'default' && !parentPath.parentPath.isProgram()) {
    parentPath = parentPath.parentPath
  }

  const css = cssToBabelType(styles.css)
  const newPath = t.isArrayExpression(css)
    ? css
    : t.newExpression(t.identifier('String'), [css])

  // default exports

  if (baseExportName === 'default') {
    const defaultExportIdentifier = path.scope.generateUidIdentifier(
      'defaultExport'
    )
    parentPath.insertBefore(
      t.variableDeclaration('const', [
        t.variableDeclarator(defaultExportIdentifier, newPath)
      ])
    )

    parentPath.insertBefore(addHash(defaultExportIdentifier, styles.hash))
    path.replaceWith(defaultExportIdentifier)
    return
  }

  // local and named exports

  parentPath.insertAfter(addHash(t.identifier(baseExportName), styles.hash))
  path.replaceWith(newPath)
}

function addHash(exportIdentifier, hash) {
  const value = typeof hash === 'string' ? t.stringLiteral(hash) : hash
  return t.expressionStatement(
    t.assignmentExpression(
      '=',
      t.memberExpression(exportIdentifier, t.identifier('__hash')),
      value
    )
  )
}

export const visitor = {
  ImportDeclaration(path, state) {
    // import css from 'styled-jsx/css'
    if (path.node.source.value !== 'styled-jsx/css') {
      return
    }

    // Find all the imported specifiers.
    // e.g import css, { global, resolve } from 'styled-jsx/css'
    // -> ['css', 'global', 'resolve']
    const specifiersNames = path.node.specifiers.map(
      specifier => specifier.local.name
    )
    specifiersNames.forEach(tagName => {
      // Get all the reference paths i.e. the places that use the tagName above
      // eg.
      // css`div { color: red }`
      // css.global`div { color: red }`
      // global`div { color: red `
      const binding = path.scope.getBinding(tagName)

      if (!binding || !Array.isArray(binding.referencePaths)) {
        return
      }

      // Produces an object containing all the TaggedTemplateExpression paths detected.
      // The object contains { scoped, global, resolve }
      const taggedTemplateExpressions = binding.referencePaths
        .map(ref => ref.parentPath)
        .reduce(
          (result, path) => {
            let taggedTemplateExpression
            if (path.isTaggedTemplateExpression()) {
              // css`` global`` resolve``
              taggedTemplateExpression = path
            } else if (
              path.parentPath &&
              path.isMemberExpression() &&
              path.parentPath.isTaggedTemplateExpression()
            ) {
              // This part is for css.global`` or css.resolve``
              // using the default import css
              taggedTemplateExpression = path.parentPath
            } else {
              return result
            }

            const tag = taggedTemplateExpression.get('tag')
            const id = tag.isIdentifier()
              ? tag.node.name
              : tag.get('property').node.name

            if (result[id]) {
              result[id].push(taggedTemplateExpression)
            } else {
              result.scoped.push(taggedTemplateExpression)
            }
            return result
          },
          {
            scoped: [],
            global: [],
            resolve: []
          }
        )

      let hasJSXStyle = false

      const { vendorPrefix, sourceMaps } = state.opts

      Object.keys(taggedTemplateExpressions).forEach(type =>
        taggedTemplateExpressions[type].forEach(path => {
          hasJSXStyle = true
          // Process each css block
          processTaggedTemplateExpression({
            type,
            path,
            fileInfo: {
              file: state.file,
              sourceFileName: state.file.opts.sourceFileName,
              sourceMaps
            },
            splitRules:
              typeof state.opts.optimizeForSpeed === 'boolean'
                ? state.opts.optimizeForSpeed
                : process.env.NODE_ENV === 'production',
            plugins: state.plugins,
            vendorPrefix
          })
        })
      )

      // When using the `resolve` helper we need to add an import
      // for the _JSXStyle component `styled-jsx/style`
      if (
        hasJSXStyle &&
        taggedTemplateExpressions.resolve.length > 0 &&
        !state.hasInjectedJSXStyle &&
        !path.scope.hasBinding(STYLE_COMPONENT)
      ) {
        state.hasInjectedJSXStyle = true
        const importDeclaration = createReactComponentImportDeclaration()
        path.scope.path.node.body.unshift(importDeclaration)
      }
    })

    // Finally remove the import
    path.remove()
  }
}

export default function() {
  return {
    Program(path, state) {
      setStateOptions(state)
    },
    ...visitor
  }
}
