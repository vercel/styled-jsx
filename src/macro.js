import { createMacro, MacroError } from 'babel-plugin-macros'
import { processTaggedTemplateExpression } from './babel-external'
import {
  setStateOptions,
  createReactComponentImportDeclaration
} from './_utils'
import { STYLE_COMPONENT } from './_constants'

export default createMacro(styledJsxMacro)

function styledJsxMacro({ references, state }) {
  setStateOptions(state)

  // Holds a reference to all the lines where strings are tagged using the `css` tag name.
  // We print a warning at the end of the macro in case there is any reference to css,
  // because `css` is generally used as default import name for 'styled-jsx/css'.
  // People who want to migrate from this macro to pure styled-jsx might have name conflicts issues.
  const cssReferences = []

  // references looks like this
  // {
  //    default: [path, path],
  //    resolve: [path],
  // }
  Object.keys(references).forEach(refName => {
    // Enforce `resolve` as named import so people
    // can only import { resolve } from 'styled-jsx/macro'
    // or an alias of it eg. { resolve as foo }
    if (refName !== 'default' && refName !== 'resolve') {
      throw new MacroError(
        `Imported an invalid named import: ${refName}. Please import: resolve`
      )
    }

    // Start processing the references for refName
    references[refName].forEach(path => {
      // We grab the parent path. Eg.
      // path -> css
      // path.parenPath -> css`div { color: red }`
      let templateExpression = path.parentPath

      // templateExpression member expression?
      // path -> css
      // path.parentPath -> css.resolve
      if (templateExpression.isMemberExpression()) {
        // grab .resolve
        const tagPropertyName = templateExpression.get('property').node.name
        // Member expressions are only valid on default imports
        // eg. import css from 'styled-jsx/macro'
        if (refName !== 'default') {
          throw new MacroError(
            `Can't use named import ${path.node.name} as a member expression: ${
              path.node.name
            }.${tagPropertyName}\`div { color: red }\` Please use it directly: ${
              path.node.name
            }\`div { color: red }\``
          )
        }
        // Otherwise enforce `css.resolve`
        if (tagPropertyName !== 'resolve') {
          throw new MacroError(
            `Using an invalid tag: ${tagPropertyName}. Please use ${
              templateExpression.get('object').node.name
            }.resolve`
          )
        }

        // Grab the TaggedTemplateExpression
        // i.e. css.resolve`div { color: red }`
        templateExpression = templateExpression.parentPath
      } else {
        if (refName === 'default') {
          const { name } = path.node
          throw new MacroError(
            `Can't use default import directly eg. ${name}\`div { color: red }\`. Please use ${name}.resolve\`div { color: red }\` instead.`
          )
        }

        if (path.node.name === 'css') {
          // If the path node name is `css` we push it to the references above to emit a warning later.
          cssReferences.push(path.node.loc.start.line)
        }
      }

      // Finally transform the path :)
      processTaggedTemplateExpression({
        type: 'resolve',
        path: templateExpression,
        fileInfo: {
          file: state.file,
          sourceFileName: state.file.opts.sourceFileName,
          sourceMaps: state.opts.sourceMaps
        },
        splitRules:
          typeof state.opts.optimizeForSpeed === 'boolean'
            ? state.opts.optimizeForSpeed
            : process.env.NODE_ENV === 'production',
        plugins: state.plugins,
        vendorPrefix: state.opts.vendorPrefixes
      })

      if (
        !state.hasInjectedJSXStyle &&
        !path.scope.hasBinding(STYLE_COMPONENT)
      ) {
        state.hasInjectedJSXStyle = true
        const importDeclaration = createReactComponentImportDeclaration()
        path.findParent(p => p.isProgram()).node.body.unshift(importDeclaration)
      }
    })
  })

  if (cssReferences.length > 0) {
    console.warn(
      `styled-jsx - Warning - We detected that you named your tag as \`css\` at lines: ${cssReferences.join(
        ', '
      )}.\n` +
        'This tag name is usually used as default import name for `styled-jsx/css`.\n' +
        'Porting macro code to pure styled-jsx in the future might be a bit problematic.'
    )
  }
}
