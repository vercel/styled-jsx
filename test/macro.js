// Packages
import test from 'ava'

// Ours
import macros from 'babel-plugin-macros'
import jsx from 'babel-plugin-syntax-jsx'
import _transform, { transformSource as _transformSource } from './_transform'

const transform = (file, opts = {}) =>
  _transform(file, {
    plugins: [macros, jsx],
    ...opts
  })

const transformSource = (src, opts = {}) =>
  _transformSource(src.trim(), {
    plugins: [macros, jsx],
    ...opts
  })

test('transpiles correctly', async t => {
  const { code } = await transform('./fixtures/macro.js')
  t.snapshot(code)
})

test('throws when using the default export directly', async t => {
  const { message } = await t.throws(
    transformSource(`
    import css from './src/macro'

    css\`div { color: red }\`
  `)
  )

  t.regex(message, /can't use default import directly/i)
})

test('throws when using the default export directly and it is not called css', async t => {
  const { message } = await t.throws(
    transformSource(`
    import foo from './src/macro'

    foo\`div { color: red }\`
  `)
  )

  t.regex(message, /can't use default import directly/i)
})

test('throws when using the default export directly and it is not called resolve', async t => {
  const { message } = await t.throws(
    transformSource(`
    import resolve from './src/macro'

    resolve\`div { color: red }\`
  `)
  )

  t.regex(message, /can't use default import directly/i)
})

test('throws when using an invalid method from the default export', async t => {
  const { message } = await t.throws(
    transformSource(`
    import css from './src/macro'

    css.foo\`div { color: red }\`
  `)
  )

  t.regex(message, /using an invalid tag/i)
})

test('throws when using a named import different than resolve', async t => {
  const { message } = await t.throws(
    transformSource(`
    import { foo } from './src/macro'

    foo\`div { color: red }\`
  `)
  )

  t.regex(message, /imported an invalid named import/i)
})

test('throws when using a named import as a member expression', async t => {
  const { message } = await t.throws(
    transformSource(`
    import { resolve } from './src/macro'

    resolve.foo\`div { color: red }\`
  `)
  )

  t.regex(message, /can't use named import/i)
})

test('can alias the named import', async t => {
  const { code } = await transformSource(`
    import { resolve as foo } from './src/macro'

    foo\`div { color: red }\`
  `)
  t.snapshot(code)
})

test('injects JSXStyle for nested scope', async t => {
  const { code } = await transformSource(`
    import { resolve } from './src/macro'

    function test() {
      resolve\`div { color: red }\`
    }
  `)
  t.snapshot(code)
})
