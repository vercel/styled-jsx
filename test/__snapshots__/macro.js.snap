// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can alias the named import 1`] = `
"import _JSXStyle from 'styled-jsx/style';


({
    styles: <_JSXStyle styleId={\\"2886504620\\"} css={\\"div.jsx-2886504620{color:red;}\\"} />,
    className: 'jsx-2886504620'
});"
`;

exports[`injects JSXStyle for nested scope 1`] = `
"import _JSXStyle from 'styled-jsx/style';


function test() {
  ({
    styles: <_JSXStyle styleId={\\"2886504620\\"} css={\\"div.jsx-2886504620{color:red;}\\"} />,
    className: 'jsx-2886504620'
  });
}"
`;

exports[`transpiles correctly 1`] = `
"import _JSXStyle from 'styled-jsx/style';


const { className, styles } = {
  styles: <_JSXStyle styleId={\\"2052294191\\"} css={\\"div.jsx-2052294191{color:red;}\\"} />,
  className: 'jsx-2052294191'
};

const dynamicStyles = props => ({
  styles: <_JSXStyle styleId={\\"290194820\\"} css={\`div.__jsx-style-dynamic-selector{color:\${props.color};}\`} dynamic={[props.color]} />,
  className: _JSXStyle.dynamic([['290194820', [props.color]]])
});

const test = {
  styles: <_JSXStyle styleId={\\"2052294191\\"} css={\\"div.jsx-2052294191{color:red;}\\"} />,
  className: 'jsx-2052294191'
};

const dynamicStyles2 = props => ({
  styles: <_JSXStyle styleId={\\"290194820\\"} css={\`div.__jsx-style-dynamic-selector{color:\${props.color};}\`} dynamic={[props.color]} />,
  className: _JSXStyle.dynamic([['290194820', [props.color]]])
});

const ExampleComponent = props => {
  const { className, styles } = dynamicStyles(props);

  return <div className={className}>
      howdy
      {styles}
    </div>;
};"
`;
